trinity/chains/light.py

```diff
@@ -0,0 +1,230 @@
+import asyncio
+import inspect
+from typing import (  # noqa: F401
+    Any,
+    Optional,
+    Callable,
+    cast,
+    Dict,
+    Generator,
+    Iterator,
+    Tuple,
+    Type,
+    TYPE_CHECKING,
+    Union,
+)
+
+from eth_typing import (
+    BlockNumber,
+    Hash32,
+)
+
+from evm.chains.base import (
+    AccountState,
+    BaseChain,
+)
+from evm.db.backends.base import BaseDB
+from evm.db.chain import (
+    BaseChainDB,
+)
+from evm.db.header import (
+    BaseHeaderDB,
+)
+from evm.exceptions import (
+    HeaderNotFound,
+    VMNotFound,
+)
+from evm.rlp.blocks import (
+    BaseBlock,
+)
+from evm.rlp.headers import (
+    BlockHeader,
+    HeaderParams,
+)
+from evm.rlp.transactions import (
+    BaseTransaction,
+    BaseUnsignedTransaction,
+)
+from evm.validation import (
+    validate_block_number,
+)
+
+from p2p.lightchain import (
+    LightPeerChain,
+)
+
+if TYPE_CHECKING:
+    from evm.vm.base import BaseVM  # noqa: F401
+
+
+class LightDispatchChain(BaseChain):
+    """
+    Provide the :class:`BaseChain` API, even though only a
+    :class:`LightPeerChain` is syncing. Store results locally so that not
+    all requests hit the light peer network.
+    """
+
+    ASYNC_TIMEOUT_SECONDS = 10
+    _loop = None
+
+    def __init__(self, headerdb: BaseHeaderDB, peer_chain: LightPeerChain) -> None:
+        self._headerdb = headerdb
+        self._peer_chain = peer_chain
+        self._peer_chain_loop = asyncio.get_event_loop()
+
+    #
+    # Helpers
+    #
+    @classmethod
+    def get_chaindb_class(cls) -> Type[BaseChainDB]:
+        raise NotImplementedError("Chain classes must implement " + inspect.stack()[0][3])
+
+    #
+    # Chain API
+    #
+    @classmethod
+    def from_genesis(cls,
+                     base_db: BaseDB,
+                     genesis_params: Dict[str, HeaderParams],
+                     genesis_state: AccountState=None) -> 'BaseChain':
+        raise NotImplementedError("Chain classes must implement " + inspect.stack()[0][3])
+
+    @classmethod
+    def from_genesis_header(cls,
+                            base_db: BaseDB,
+                            genesis_header: BlockHeader) -> 'BaseChain':
+        raise NotImplementedError("Chain classes must implement " + inspect.stack()[0][3])
+
+    def get_chain_at_block_parent(self, block: BaseBlock) -> 'BaseChain':
+        raise NotImplementedError("Chain classes must implement " + inspect.stack()[0][3])
```

Comment:
Shouldn't the method name be available as part of the stacktrace?

Reply:
When RPC fails, it only returns the exception message. It was easier to see it in the web3py failure message than go hunting for the stack trace in the trinity logs.